pkg/service/controller/servingcert/controller.go

```diff
@@ -0,0 +1,239 @@
+	queue      workqueue.RateLimitingInterface
+	maxRetries int
+
+	serviceCache      cache.Store
+	serviceController *framework.Controller
+
+	ca         *crypto.CA
+	publicCert string
+	dnsSuffix  string
+
+	// syncHandler does the work. It's factored out for unit testing
+	syncHandler func(serviceKey string) error
+}
+
+// NewServiceServingCertController creates a new ServiceServingCertController.
+func NewServiceServingCertController(serviceClient kclient.ServicesNamespacer, secretClient kclient.SecretsNamespacer, ca *crypto.CA, dnsSuffix string, resyncInterval time.Duration) *ServiceServingCertController {
+	sc := &ServiceServingCertController{
+		serviceClient: serviceClient,
+		secretClient:  secretClient,
+
+		queue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		maxRetries: 10,
+
+		ca:        ca,
+		dnsSuffix: dnsSuffix,
+	}
+
+	sc.serviceCache, sc.serviceController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
+				return sc.serviceClient.Services(kapi.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
+				return sc.serviceClient.Services(kapi.NamespaceAll).Watch(options)
+			},
+		},
+		&kapi.Service{},
+		resyncInterval,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				service := obj.(*kapi.Service)
+				glog.V(4).Infof("Adding service %s", service.Name)
+				sc.enqueueService(obj)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				service := cur.(*kapi.Service)
+				glog.V(4).Infof("Updating service %s", service.Name)
+				// Resync on service object relist.
+				sc.enqueueService(cur)
+			},
+		},
+	)
+
+	sc.syncHandler = sc.syncService
+
+	return sc
+}
+
+// Run begins watching and syncing.
+func (sc *ServiceServingCertController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	go sc.serviceController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go wait.Until(sc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down service signing cert controller")
+	sc.queue.ShutDown()
+}
+
+func (sc *ServiceServingCertController) enqueueService(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+
+	sc.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (sc *ServiceServingCertController) worker() {
+	for {
+		func() {
+			key, quit := sc.queue.Get()
+			if quit {
+				return
+			}
+			defer sc.queue.Done(key)
+			err := sc.syncHandler(key.(string))
+			if err != nil {
+				if sc.queue.NumRequeues(key) >= sc.maxRetries {
+					utilruntime.HandleError(fmt.Errorf("Error syncing service, it will not be retried: %v", err))
```

Comment:
caps vs not.

Also, technically, it will be retried on the next sync interval.  We're probably going to have to set our sync to ~ a few minutes to ensure we handle errors like this (or time the errors out more)